doc/manual/syntax.md

```diff
@@ -619,19 +648,182 @@ error: contract broken by a value.
-    and is accurate and reliable.
-    (Collins dictionary)
-    "%
+> {foo = [[1]]} : {foo : Array (Array Number)}
+{ foo = [ [ 1 ] ] }
+
+> let select
+    : forall a. {left: a, right: a} -> [| `left, `right |] -> a
+    = fun {left, right} =>
+       match {
+         `left => left,
+         `right => right,
+       }
+  in
+  (select {left = true, right = false} `left) : Bool
 true
+
+> let add_foo : forall a. {_: a} -> a -> {_: a} = fun dict value =>
+    record.insert "foo" value dict
+in
+add_foo {bar = 1} 5 : _
+{ bar = 1, foo = 5 }
+
+> {foo = 1, bar = "string"} : {_ : Number}
+error: incompatible types
+  ┌─ repl-input-12:1:17
+  │
+1 │ {foo = 1, bar = "string"} : {_ : Number}
+  │                 ^^^^^^^^ this expression
+  │
+  = The type of the expression was expected to be `Number`
+  = The type of the expression was inferred to be `String`
+  = These types are not compatible
+```
+
+#### Record types
+
+Record types are syntactically a restricted subset of records literal. They are
+handled differently from normal record literals with respect to typechecking.
+
+A record literal is a record type if:
+
+- No field has a defined value: there are only fields without a definition.
+- Each field has exactly one type annotation
+- Each field doesn't have any other metadata attached (see [Metadata](#metadata))
+- The record literal might have a record tail, written as
+  `{ <fields> ; <tail> }`. The tail appears at the end of the field declarations
+  and is preceded by `;`. `<tail>` must be a valid identifier.
+
+If the above properties are satisfied, a record literal is considered to be a
+record type by the typechecker.
+
+Trying to attach a tail `; tail` to a record literal which isn't a record type
+is a parse error.
+
+Examples:
+
+```text
+> {foo = 1, bar = "foo" } : {foo : Number, bar: String}
+{ bar = "foo", foo = 1 }
+
+> {foo.bar = 1, baz = 2} : {foo: {bar : Number}, baz : Number}
+{ baz = 2, foo = { bar = 1 } }
+
+># the right-hand side is missing a type annotation for baz, so it isn't a
+ # record type
+ {foo = 1, bar = "foo" } : {foo : Number, bar : String, baz}
+error: incompatible types
+  ┌─ repl-input-6:1:1
+  │
+1 │ {foo = 1, bar = "foo" } : {foo : Number, bar : String, baz}
+  │ ^^^^^^^^^^^^^^^^^^^^^^^ this expression
+  │
+  = The type of the expression was expected to be `{ bar : String, baz, foo : Numb…` (a contract)
+  = The type of the expression was inferred to be `{bar: _a, foo: _b}`
+  = Static types and contracts are not compatible
+
+> # the annotation has a metadata
+ {foo = 1, bar = "foo" } : {foo : Number, bar : String | optional}
+error: incompatible types
+[..]
+
+> # while MyDyn isn't a proper type, the right record respect all the
+  # requirements of a record type
+let MyDyn = fun label value => value in
+{foo = 1, bar | MyDyn = "foo"} : {foo : Number, bar : MyDyn}
+{ bar = "foo", foo = 1 }
+```
+
+## Metadata
+
+Metadata are used to attach type and contract annotations, documentation, a
```

Comment:
```suggestion
Metadata is used to attach type and contract annotations, documentation, a
```

Reply:
Waiting for the conclusion of the "plural vs singular" metadata for this one.